htlcswitch/payment_result.go

```diff
@@ -0,0 +1,268 @@
+package htlcswitch
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+
+	// paymentStoreBucketKey is used for the root level bucket that
+	// stores the payment result for each payment ID.
+	paymentStoreBucketKey = []byte("payment-result-store-bucket")
+
+	// ErrPaymentIDNotFound is an error returned if the given paymentID is
+	// not found.
+	ErrPaymentIDNotFound = errors.New("paymentID not found")
+
+	// ErrPaymentIDAlreadyExists is returned if we try to write a pending
+	// payment whose paymentID already exists.
+	ErrPaymentIDAlreadyExists = errors.New("paymentID already exists")
+)
+
+// PaymentResult wraps a result received from the network after a payment
+// attempt was made.
+type PaymentResult struct {
+	// Preimage is set by the switch in case a sent HTLC was settled.
+	Preimage [32]byte
+
+	// Error is non-nil in case a HTLC send failed, and the HTLC is now
+	// irrevocably cancelled.
+	Error *ForwardingError
+}
+
+// serializePaymentResult serializes the PaymentResult.
+func serializePaymentResult(w io.Writer, p *PaymentResult) error {
+	if err := channeldb.WriteElements(w,
+		p.Preimage[:],
+	); err != nil {
+		return err
+	}
+
+	// If there's no error, write 'false' and return.
+	if p.Error == nil {
+		return channeldb.WriteElements(w, false)
+	}
+
+	// Otherwise, write 'true' and the error fields.
+	msg := []byte(p.Error.ExtraMsg)
+	err := channeldb.WriteElements(w, true,
+		p.Error.ErrorSource, msg,
```

Comment:
`ErrorSource` could be `nil` if the error source is unknown (for example after restart or when the onion decrypt fails). The store should probably prepare for that, to prevent a db migration.

Reply:
Using `nil` can leads to problems down the line, what do you think about defining a constant `UnknownPaymentSource`?

Reply:
What kind of problems? `nil` for unset fields is quite common? For serialize to the db we can of course write anything we want.